File: s3-source-connector/src/main/java/sourceconnector/service/processor/BaseProcessor.java
package sourceconnector.service.processor;

public abstract class BaseProcessor<T> implements Processor<T, T> {
  protected BaseProcessor<T> next;

  public BaseProcessor<T> setNext(BaseProcessor<T> nextProcessor) {
    this.next = nextProcessor;
    return nextProcessor;
  }

  /**
   * Process input and return the final output
   */
  public abstract T process(T input);
}

File: source-connector/src/main/kotlin/sourceconnector/domain/processor/BaseProcessor.kt
package sourceconnector.domain.processor

abstract class BaseProcessor<T> : Processor<T, T?> {
  protected var next: BaseProcessor<T>? = null

  fun setNext(nextProcessor: BaseProcessor<T>): BaseProcessor<T> {
    this.next = nextProcessor
    return nextProcessor
  }

  /**
   * Process input and return the final output
   */
  abstract override fun process(input: T): T?
}
